src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -178,21 +179,35 @@ public bool TrySetRunning()
                 return true;
             }
 
-            public void SetComplete()
+            public bool TrySetComplete()
             {
-                Debug.Assert(Volatile.Read(ref _state) == (int)State.Running);
+                State oldState = (State)Interlocked.CompareExchange(ref _state, (int)State.Complete, (int)State.Running);
+                if (oldState == State.Aborted)
```

Comment:
suggestion: `return oldState == State.Running`.

Reply:
and maybe before that `Debug.Assert(oldState == State.Running || oldState == State.Aborted)`.

Reply:
`TrySetRunning` needs to be updated also.

Reply:
@tmds  this looks better, i'll make the changes thanks.

For `TrySetRunning`, it seems that changes are not required there if we always only transition from `Running` to `Aborted`. However, if we can also transition from `Waiting` to `Aborted`, then it would need to be changed.

When writing the fix, I tried to define `Aborted` and `Cancelled` as follows which sounds consistent. It might need some discussion:
`Aborted` - an operation is (ungracefully) stopped during its execution
`Cancelled` - an operation is stopped before its even executed or it's idle waiting (which means it's been gracefully stopped)

what do you think?

Reply:
@tmds i've made the changes, also allowing transitions from any state to `Aborted`